packages/utils/telemetry-utils/src/errorLogging.ts

```diff
@@ -3,102 +3,202 @@
+ * If the object is not suitable for annotation (e.g. it's frozen), this will throw.
+ * @param error - The error to annotate by patching in the properties of IFluidErrorBase
+ * @param annotations - Annotations to apply to the normalized error
  */
-type RwLoggingError = LoggingError;
+export function annotateErrorObject(
+    errorObject: unknown,
+    annotations: FluidErrorAnnotations = {},
+): asserts errorObject is IFluidErrorBase {
+    // If errorObject isn't suitable to annotate as IFluidErrorBase, bail.
+    assert(isRegularObject(errorObject) && !Object.isFrozen(errorObject),
+        "Cannot annotate a non-object or frozen error");
+
+    // Patch in any missing properties of IFluidErrorBase
+    const errorTypeIfNone = isErrorLike(errorObject)
+        ? errorObject.name
+        : typeof(errorObject);
+    const fluidError = patchFluidErrorBuilder(
+        errorObject as FluidErrorBuilder,
+        prepareFluidErrorTemplate(errorObject, errorTypeIfNone, annotations.errorCodeIfNone),
+    );
 
-/** Type guard for RwLoggingError.  Not exported. */
-const isRwLoggingError = (x: any): x is RwLoggingError =>
-    typeof x?.addTelemetryProperties === "function" && isILoggingError(x);
+    mixinTelemetryPropsWithFluidError(fluidError, annotations.props);
+}
 
 /**
- * Annotate the given error object with the given logging props
- * @returns The same error object passed in if possible, with telemetry props functionality mixed in
+ * Helper type, not exported.
+ * Makes IFluidErrorBuilder's props mutable and optional for building one up on an existing object
  */
-export function annotateError(
-    error: unknown,
-    props: ITelemetryProperties,
-): ILoggingError {
-    if (isRwLoggingError(error)) {
-        error.addTelemetryProperties(props);
-        return error;
-    }
+type FluidErrorBuilder = {
+    -readonly [P in keyof IFluidErrorBase]?: IFluidErrorBase[P];
+};
+
+/**
+ * Helper type, not exported.
+ * Makes IFluidErrorBuilder's props all type unknown for querying if they're already valid
+ */
+ type FluidErrorProperties = {
+    -readonly [P in keyof IFluidErrorBase]: unknown;
+};
 
-    if (isRegularObject(error)) {
-        // Even though it's not exposed, fully implement IRwLoggingError for subsequent calls to annotateError
-        const loggingError = error as RwLoggingError;
+/** Returns a template IFluidErrorBase with values based on the inputs provided, for use with patchFluidErrorBuilder */
+function prepareFluidErrorTemplate(
+    originalError: unknown,
+    errorTypeIfNone: string,
+    errorCodeIfNone: string | undefined,
+): IFluidErrorBase {
+    const fullErrorTypeIfNone = `none (${errorTypeIfNone})`;
+    const fullErrorCodeIfNone = errorCodeIfNone === undefined
+        ? "none"
+        : `none (${errorCodeIfNone})`;
+    const { errorType, fluidErrorCode, message, name, stack } = originalError as FluidErrorProperties;
```

Comment:
This assumes that originalError is an object. Is that valid assumption? If so, I'd suggest to change type of originalError to reflect that.

Reply:
Ha I spent some time trying to figure out what type can be used for this kind of constraint, and there's not a great answer.  e.g. previous `mixinTelemetryProps` had a type constraint but I ditched it for simplicity.  I guess I'll add an assert, although it's a bit excessive since this is a "private" helper.

Reply:
Actually I can use `Object` since I'm just destructuring.